timemachine/fe/dummy.py

```diff
@@ -0,0 +1,396 @@
+import numpy as np
+import networkx as nx
+import copy
+import itertools
+
+
+def _add_successors(mol, core, groups):
+    # (ytz): internal utility used for bfs
+    next_groups = []
+    for group in groups:
+        # grab the last node
+        last_node = group[-1]
+        for nbr in mol.GetAtomWithIdx(last_node).GetNeighbors():
+            new_group = copy.deepcopy(group)
+            nbr = nbr.GetIdx()
+            # ensure is core_atom and not already visited
+            if nbr in core and nbr not in new_group:
+                new_group.append(nbr)
+                next_groups.append(new_group)
+
+    return next_groups
+
+
+def identify_anchor_groups(mol, core, anchor):
+    """
+    Generate putative choices for valid anchor groups. An anchor group
+    is an ordered sequence of core atoms starting with the anchor (a-b-c) that
+    are connected by bonds. In other words, an anchor group is a rooted subtree
+    with three nodes spanning from the anchor.
+
+    Parameters
+    ----------
+    mol: Chem.Mol
+        rdkit molecule
+
+    core: list or set or iterable
+        core atoms
+
+    dummy_atom: int
+        atom we're initializing the search over
+
+    Returns
+    -------
+    3-tuple
+        Returns anchor groups of size 1, 2, and 3. Size 1 group will always
+        be [[anchor]]. Size 2 groups will be [[anchor, x], ...] and size 3
+        groups will be [[anchor,x,y,], ...]
+
+    """
+
+    assert anchor in core
+
+    # We perform a bfs of depth 3 starting from the anchor.
+    layer_1_groups = [[anchor]]
+    layer_2_groups = _add_successors(mol, core, layer_1_groups)
+    layer_3_groups = _add_successors(mol, core, layer_2_groups)
```

Comment:
Clarification: The reason this stops at layer_3 is because each term in our valence forcefield only "talks to" a group of atoms up to 3 bonds away, correct?

------

Condition where this could be invalidated in the future: if the force field ever defines improper torsions on sets of 4 atoms (a, b, c, d) where shortest_path_length(a, d) > 3.

Defensive request: If the depth=3 here relies on this sort of assumption, could we add a function to check that a bond table + set of PeriodicTorsion indices are consistent with the assumption?

Reply:
> Clarification: The reason this stops at layer_3 is because each term in our valence forcefield only "talks to" a group of atoms up to 3 bonds away, correct?

Pretty much - with the extra condition that 1-2 interactions must be defined over bonds, 1-3 over angles, and 1-4 over torsions. 

> Condition where this could be invalidated in the future: if the force field ever defines improper torsions on sets of 4 atoms (a, b, c, d) where shortest_path_length(a, d) > 3.

Impropers are technically fine from a correctness point of view. Underneath the hood they're just implemented as proper torsions of 4 indices. The difference lies in which bonds/angles are involved. A typical ProperTorsion has 1 torsion, 2 angle, and 3 bonded terms. An Improper Torsion has 1 torsion, 1 angle, and 2 bonded terms. We may run the risk of including a torsion without all the relevant bond/angle terms, but from a technically correct standpoint you can probably still factorize the resulting partition function (eg. in 2D you can define only an angle restraint without adding the bond restraints, the partition function still factorizes, but it just may be very hard to converge since the dummy atom will be drifting away). 
 
One good check we could assert (not here but at a later stage) is that every 1-3 term has both 1-2 terms turned on, and every 1-4 term has both 1-3 terms and all three 1-2 terms turned on.

Reply:
`strict` mode will now check that that every 1-3 term has both 1-2 terms turned on, and every 1-4 term has both 1-3 terms and all three 1-2 terms turned on.